File: back-end/src/main/java/es/ull/project/domain/valueobject/location/Location.java
package es.ull.project.domain.valueobject.location;

import es.ull.utils.geolocation.UllGeolocationPoint;

import java.util.Objects;
import java.util.regex.Pattern;

/**
 * Location
 * 
 * Represents a physical location.
 * Can contain coordinates (latitude, longitude), postal address, and GIS reference.
 * This is a mandatory value object in the domain.
 */
public final class Location {

    /**
     * Postal address constraints.
     */
    private static final int POSTAL_ADDRESS_MAX_LENGTH = 150;
    private static final Pattern POSTAL_ADDRESS_PATTERN = Pattern.compile("^[A-Za-z0-9\\s,.-]+$");
    private static final String ERROR_ADDRESS_NOT_DEFINED = "Postal address is not defined";
    private static final String ERROR_ADDRESS_EMPTY = "Postal address cannot be empty";
    private static final String ERROR_ADDRESS_MAX_LENGTH = "Postal address must be at most " + POSTAL_ADDRESS_MAX_LENGTH + " characters";
    private static final String ERROR_ADDRESS_FORMAT = "Postal address format is invalid";

    /**
     * GIS reference constraints.
     */
    private static final int GIS_REFERENCE_MAX_LENGTH = 100;
    private static final String ERROR_GIS_NOT_DEFINED = "GIS reference is not defined";
    private static final String ERROR_GIS_EMPTY = "GIS reference cannot be empty";
    private static final String ERROR_GIS_MAX_LENGTH = "GIS reference must be at most " + GIS_REFERENCE_MAX_LENGTH + " characters";

    /**
     * Zero constant for comparisons.
     * Used to avoid magic numbers in the code and improve readability.
     */
    private static final int ZERO = 0;

    /**
     * Geolocation point containing latitude and longitude.
     * It is a required attribute.
     */
    private final UllGeolocationPoint point;

    /**
     * Postal address.
     * It is a required attribute.
     */
    private final String postalAddress;

    /**
     * GIS reference. (Geographic Information System)
     * TODO: maybe change String to a more complex type if needed in the future.
     * It is a required attribute.
     */
    private final String gisReference;

    /**
     * Creates a new Location with coordinates, postal address, and GIS reference.
     * 
     * @param latitude Latitude in decimal degrees.
     * @param longitude Longitude in decimal degrees.
     * @param postalAddress Postal address.
     * @param gisReference GIS reference. (Geographic Information System)
     */
    public Location(double latitude, double longitude, String postalAddress, String gisReference) {
        this.point = new UllGeolocationPoint(longitude, latitude);
        this.validatePostalAddress(postalAddress);
        this.validateGISReference(gisReference);
        this.postalAddress = postalAddress;
        this.gisReference = gisReference;
    }

    /**
     * Validates the postal address.
     *
     * @param postalAddress the postal address to validate
     * @throws IllegalArgumentException if the postal address is invalid
     */
    private void validatePostalAddress(String postalAddress) {
        if (postalAddress == null) {
            throw new IllegalArgumentException(ERROR_ADDRESS_NOT_DEFINED);
        }
        int length = postalAddress.length();
        if (length == ZERO) {
            throw new IllegalArgumentException(ERROR_ADDRESS_EMPTY);
        }
        if (length > POSTAL_ADDRESS_MAX_LENGTH) {
            throw new IllegalArgumentException(ERROR_ADDRESS_MAX_LENGTH);
        }
        if (!POSTAL_ADDRESS_PATTERN.matcher(postalAddress).matches()) {
            throw new IllegalArgumentException(ERROR_ADDRESS_FORMAT);
        }
    }

    /**
     * Validates the GIS reference.
     *
     * @param gisReference the GIS reference to validate
     * @throws IllegalArgumentException if the GIS reference is invalid
     */
    private void validateGISReference(String gisReference) {
        if (gisReference == null) {
            throw new IllegalArgumentException(ERROR_GIS_NOT_DEFINED);
        }
        int length = gisReference.length();
        if (length == ZERO) {
            throw new IllegalArgumentException(ERROR_GIS_EMPTY);
        }
        if (length > GIS_REFERENCE_MAX_LENGTH) {
            throw new IllegalArgumentException(ERROR_GIS_MAX_LENGTH);
        }
    }

    /**
     * Returns the geolocation point.
     *
     * @return the geolocation point
     */
    public UllGeolocationPoint getPoint() {
        return point;
    }

    /**
     * Returns the latitude in decimal degrees.
     *
     * @return the latitude
     */
    public double getLatitude() {
        return point.getLatitude();
    }

    /**
     * Returns the longitude in decimal degrees.
     *
     * @return the longitude
     */
    public double getLongitude() {
        return point.getLongitude();
    }

    /**
     * Returns the postal address.
     *
     * @return the postal address
     */
    public String getPostalAddress() {
        return postalAddress;
    }

    /**
     * Returns the GIS reference.
     *
     * @return the GIS reference
     */
    public String getGISReference() {
        return gisReference;
    }

    /**
     * Compares this location with another object for equality.
     *
     * @param otherObject the object to compare with
     * @return true if the objects are equal, false otherwise
     */
    @Override
    public boolean equals(Object otherObject) {
        if (this == otherObject) {
            return true;
        }
        if (otherObject == null || getClass() != otherObject.getClass()) {
            return false;
        }
        Location other = (Location) otherObject;
        return point.equals(other.point) &&
               postalAddress.equals(other.postalAddress) &&
               gisReference.equals(other.gisReference);
    }

    /**
     * Returns the hash code of this location.
     *
     * @return the hash code
     */
    @Override
    public int hashCode() {
        return Objects.hash(point, postalAddress, gisReference);
    }

    /**
     * Returns a string representation of this location.
     *
     * @return a formatted string with location details
     */
    @Override
    public String toString() {
        return String.format("Location={latitude=%s, longitude=%s, postalAddress='%s', gisReference='%s'}",
                point.getLatitude(), point.getLongitude(), postalAddress, gisReference);
    }
}

File: front-end/src/domain/valueobject/location/location.ts
import { UllGeolocationPoint } from '@ull-tfg/ull-tfg-typescript';

/**
 * Location
 *
 * Represents a physical location with latitude, longitude, postal address, and GIS reference.
 */
export class Location {
  /**
   * Maximum allowed length for the postal address.
   */
  private static readonly POSTAL_MAX = 150;
  /**
   * Maximum allowed length for the GIS reference.
   */
  private static readonly GIS_MAX = 100;
  /**
   * Regular expression used to validate the postal address.
   */
  private static readonly POSTAL_REGEX = /^[A-Za-z0-9\s,.-]+$/;

  /**
   * Validated postal address.
   */
  readonly postalAddress: string;
  /**
   * Validated GIS reference.
   */
  readonly gisReference: string;
  /**
   * Geolocation point used internally for validation and comparison.
   */
  private readonly point: UllGeolocationPoint;

  /**
   * Create a Location.
   * @param latitude latitude in degrees (-90..90)
   * @param longitude longitude in degrees (-180..180)
   * @param postalAddress postal address string (validated)
   * @param gisReference GIS reference string (validated)
   */
  constructor(latitude: number, longitude: number, postalAddress: string, gisReference: string) {
    this.point = new UllGeolocationPoint(longitude, latitude);
    this.validatePostalAddress(postalAddress);
    this.validateGISReference(gisReference);
    this.postalAddress = postalAddress;
    this.gisReference = gisReference;
  }

  /**
   * Latitude in degrees (-90..90).
   */
  get latitude(): number {
    return this.point.getLatitude();
  }

  /**
   * Longitude in degrees (-180..180).
   */
  get longitude(): number {
    return this.point.getLongitude();
  }

  /**
   * Validates that the postal address matches the allowed format and length.
   * @param addr Postal address to validate.
   */
  private validatePostalAddress(addr: string) {
    if (addr == null) throw new Error('Postal address is not defined');
    const length = addr.length;
    if (length === 0) throw new Error('Postal address cannot be empty');
    if (length > Location.POSTAL_MAX) throw new Error(`Postal address must be at most ${Location.POSTAL_MAX} characters`);
    if (!Location.POSTAL_REGEX.test(addr)) throw new Error('Postal address format is invalid');
  }

  /**
   * Validates that the GIS reference is not empty and does not exceed the maximum length.
   * @param gis GIS reference to validate.
   */
  private validateGISReference(gis: string) {
    if (gis == null) throw new Error('GIS reference is not defined');
    const length = gis.length;
    if (length === 0) throw new Error('GIS reference cannot be empty');
    if (length > Location.GIS_MAX) throw new Error(`GIS reference must be at most ${Location.GIS_MAX} characters`);
  }

  /**
   * Compares whether two locations are equal.
   * @param other Other object to compare.
   * @returns True if all properties match, false otherwise.
   */
  equals(other: unknown): boolean {
    if (this === other) return true;
    if (!(other instanceof Location)) return false;
    return this.point.equals(other.point) && this.postalAddress === other.postalAddress && this.gisReference === other.gisReference;
  }

  /**
   * Returns a readable representation of the location.
   * @returns String containing the location values.
   */
  toString(): string {
    return `Location={latitude=${this.latitude}, longitude=${this.longitude}, postalAddress='${this.postalAddress}', gisReference='${this.gisReference}'}`;
  }
}
